File: app/src/lib/terms.ts
/**
 * Canvas term names sort chronologically as strings — right up until Summer.
 *
 * Canvas hands back `"2026 Semester 1"`, `"2026 Semester 2"`, `"2026 Summer
 * Term"`. Compared as text, `"Summer"` beats `"Semester"` (`Su` > `Se`), so a
 * plain `ORDER BY term_name` puts Summer *last* in its year. It runs first:
 * UniMelb's Summer Term is January–February, ahead of Semester 1, with Winter
 * Term sitting between the two semesters.
 *
 * The year prefix is still fine to compare as text or as an integer; only the
 * term within a year needs a rank.
 */

/** Chronological position within an academic year. Unknown terms sort last. */
export const TERM_RANK: Record<string, number> = {
  summer: 0,
  "semester 1": 1,
  winter: 2,
  "semester 2": 3,
};

const TERM_TOKENS: ReadonlyArray<readonly [number, readonly string[]]> = [
  [0, ["summer", "sum"]],
  [1, ["semester 1", "sm1"]],
  [2, ["winter", "win"]],
  [3, ["semester 2", "sm2"]],
];

const UNKNOWN_RANK = 9;

/** `"2026 Summer Term"` → `0`. Anything unrecognised ranks after real terms. */
export function termRank(termName: string | null): number {
  if (!termName) return UNKNOWN_RANK;
  const haystack = ` ${termName.toLowerCase().split(/[\s_-]+/).join(" ")} `;
  for (const [rank, tokens] of TERM_TOKENS) {
    if (tokens.some((token) => haystack.includes(` ${token} `))) return rank;
  }
  return UNKNOWN_RANK;
}

/** `"2026 Summer Term"` → `2026`; missing or malformed years sort oldest. */
export function termYear(termName: string | null): number {
  const year = Number(termName?.slice(0, 4));
  return Number.isFinite(year) ? year : 0;
}

/** Newest term first, matching the `ORDER BY` in `getSubjects`. */
export function compareTermsNewestFirst(a: string | null, b: string | null): number {
  return termYear(b) - termYear(a) || termRank(b) - termRank(a);
}

interface TermSubject {
  id: number;
  term_name: string | null;
  workflow_state: string;
}

/** Current means the newest available term, using academic order even if an
 *  older app or a Canvas event supplied incorrect current flags. */
export function currentSubjectIds(subjects: readonly TermSubject[]): Set<number> {
  const latest = subjects
    .filter((s) => s.workflow_state === "available")
    .reduce<string | null>(
      (best, s) => compareTermsNewestFirst(s.term_name, best) < 0 ? s.term_name : best,
      null,
    );
  return new Set(subjects
    .filter((s) => s.workflow_state === "available" && compareTermsNewestFirst(s.term_name, latest) === 0)
    .map((s) => s.id));
}

/** Recognize only the obsolete automatic selection. A custom set, including
 *  selecting nothing, stays untouched. The stored flags are repaired with it,
 *  so later reads cannot repeatedly reset a person's choices. */
export function hasLegacyDefaultSelection(
  subjects: readonly (TermSubject & { is_current: boolean | number; selected: boolean | number })[],
  currentIds: ReadonlySet<number>,
): boolean {
  return subjects.some((s) => !!s.is_current)
    && subjects.every((s) => !!s.selected === !!s.is_current)
    && subjects.some((s) => !!s.is_current !== currentIds.has(s.id));
}

/**
 * The same ranking as a SQLite expression, so ordering can stay in the query
 * that fetches subjects rather than becoming a second sort in JS. Inlined as a
 * `CASE` because the plugin has no way to register a custom SQL function.
 */
export function TERM_RANK_SQL(column: string): string {
  let normalized = `replace(replace(replace(replace(replace(lower(${column}), '_', ' '), '-', ' '), char(9), ' '), char(10), ' '), char(13), ' ')`;
  for (let i = 0; i < 4; i++) normalized = `replace(${normalized}, '  ', ' ')`;
  normalized = `(' ' || ${normalized} || ' ')`;
  const whens = TERM_TOKENS
    .map(([rank, tokens]) => `WHEN ${tokens.map((token) => `${normalized} LIKE '% ${token} %'`).join(" OR ")} THEN ${rank}`)
    .join(" ");
  return `CASE ${whens} ELSE ${UNKNOWN_RANK} END`;
}

File: app/src-tauri/src/terms.rs
//! Chronological ordering for Canvas term names.
//!
//! Canvas names a term `"2026 Semester 2"` or `"2026 Summer Term"`, and those
//! strings do *not* sort chronologically: `"Summer"` beats `"Semester"` on the
//! third character (`u` > `e`), so comparing them as text files Summer last in
//! a year it actually starts. Taking the plain `.max()` of the names therefore
//! elected a single summer enrolment over the semester being studied, and
//! every real subject was stamped as past.
//!
//! Not every term is named after a semester either. A short intensive comes
//! back as the month it runs in — `"2026 June"` for `THTR30042_2026_JUN_STH_2`
//! — so the months map onto the term they fall inside: January/February are
//! Summer, June/July are Winter. Left unranked they sorted after every real
//! term, and the same `.max()` elected a one-off intensive over the semester
//! being studied.
//!
//! The year prefix compares fine either way; only the term within a year needs
//! a rank. Mirrored in `app/src/lib/terms.ts`, which the frontend uses to
//! derive the same answer at read time — both sides have to agree, because a
//! stored flag written here is what the CLI and the agent read back.

/// Chronological position within one academic year. UniMelb runs Summer Term
/// in January–February, ahead of Semester 1, with Winter Term between the two
/// semesters. An unrecognised term sorts after every real one rather than
/// silently taking a real term's place.
pub fn term_rank(name: &str) -> u8 {
    let lower = name.to_lowercase();
    let words: Vec<&str> = lower.split(|c: char| c.is_whitespace() || c == '_' || c == '-')
        .filter(|word| !word.is_empty()).collect();
    let has = |tokens: &[&str]| words.iter().any(|word| tokens.contains(word));
    let semester = |number: &str| words.windows(2).any(|pair| pair == ["semester", number]);
    if has(&["sum", "summer"]) {
        0
    } else if semester("1") || has(&["sm1"]) {
        1
    } else if has(&["win", "winter"]) {
        2
    } else if semester("2") || has(&["sm2"]) {
        3
    } else if has(&["january", "february"]) {
        0
    } else if has(&["june", "july"]) {
        2
    } else {
        9
    }
}

/// The year a term name opens with; a missing or malformed one sorts oldest.
pub fn term_year(name: &str) -> i32 {
    name.get(..4).and_then(|y| y.parse().ok()).unwrap_or(0)
}

/// Sort key putting the most recent term last, for `max_by_key`.
pub fn term_key(name: &str) -> (i32, u8) {
    (term_year(name), term_rank(name))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn summer_opens_the_year_it_names() {
        // The bug in one line: as text, Summer wins its year.
        assert!("2026 Summer Term" > "2026 Semester 2");
        // Ranked, it does not.
        assert!(term_key("2026 Summer Term") < term_key("2026 Semester 2"));
        assert!(term_key("2026 Summer Term") < term_key("2026 Semester 1"));
    }

    #[test]
    fn terms_run_in_teaching_order() {
        let mut names = vec![
            "2026 Semester 2",
            "2026 Summer Term",
            "2026 Winter Term",
            "2026 Semester 1",
        ];
        names.sort_by_key(|n| term_key(n));
        assert_eq!(
            names,
            vec![
                "2026 Summer Term",
                "2026 Semester 1",
                "2026 Winter Term",
                "2026 Semester 2",
            ]
        );
    }

    #[test]
    fn a_later_year_always_wins() {
        assert!(term_key("2026 Summer Term") > term_key("2025 Semester 2"));
    }

    #[test]
    fn the_newest_of_a_real_enrolment_is_the_semester_being_studied() {
        let terms = [
            "2024 Semester 2",
            "2025 Semester 1",
            "2026 Semester 1",
            "2026 Semester 2",
            "2026 Summer Term",
        ];
        let latest = terms.iter().max_by_key(|t| term_key(t)).unwrap();
        assert_eq!(*latest, "2026 Semester 2");
    }

    #[test]
    fn a_month_named_intensive_ranks_as_the_term_it_runs_in() {
        // The June intensive is a winter term, not an unknown one — otherwise
        // it outranks Semester 2 and steals "current" from the real enrolment.
        assert_eq!(term_rank("2026 June"), term_rank("2026 Winter Term"));
        assert!(term_key("2026 June") < term_key("2026 Semester 2"));
        assert!(term_key("2026 June") > term_key("2026 Semester 1"));
        assert_eq!(term_rank("2026 January"), term_rank("2026 Summer Term"));
        // A semester wins its own name back from the month in the brackets.
        assert_eq!(term_rank("2026 Semester 2 (July start)"), 3);
    }

    #[test]
    fn an_unknown_term_does_not_impersonate_a_real_one() {
        assert_eq!(term_rank("2026 Intensive Block"), 9);
        assert_eq!(term_year("Default Term"), 0);
    }

    #[test]
    fn codes_and_labels_share_the_same_academic_order() {
        for (code, label) in [("SUM", "Summer Term"), ("SM1", "Semester 1"),
            ("WIN", "Winter Term"), ("SM2", "Semester 2")] {
            assert_eq!(term_key(&format!("2026_{code}")), term_key(&format!("2026 {label}")));
        }
        assert_eq!(term_rank("2026-semester-2"), 3);
        assert_eq!(term_rank("2026 WINTER"), 2);
        assert_eq!(term_rank("2026 SEMESTER  1"), 1);
        assert_eq!(term_rank("2026 Semester 10"), 9);
    }
}
